### ISSUE TITLE
Create a config handler class

### ISSUE THREAD

**Issue by twVolc** [twVolc/PyCamPermanent#224]
As discussed in PR #221, and also relating to #84 somewhat, it would make sense to move all config handling to its own class (`Config`?). This would deal with all config file I/O work and setting up both `PyplisWorker` and `IFitWorker` correctly. this would tidy up `PyplisWorker` quite a bit, ensuring that it stays focussed on data processing and so is slightly less cluttered.

**Comment by ubdbra001** [comment]
Also relates to #67

**Comment by twVolc** [comment]
> * The GUI reads config values when initialised and when a new config is loaded, and writes to the config when something is changed in the GUI. The GUI will need access to the config to set the initial values for most GUI elements.

I might be missing something here, but would it be possible to have the `pyplis_worker`/`doas_worker` writing updates to the config file, rather than the GUI? That way maybe it's possible to have the GUI itself not really interacting with the config file and it's just facilitiating interactions between `pyplis_worker`/`doas_worker` and the config. That way maybe we avoid the issue of tangling too much into the GUI that doesn't need to be - in other words, if we have the GUI interacting too much with the `ConfigHandler` might we end up with a similar situation to what we have now with `PyplisWorker` and the GUI, where we wish they weren't so tangled and are now having to work to disentangle?

It's probably that I'm just struggling to picture the plans exactly - might be easier over an online discussion or with diagrams? In the end though, I'm happy to go with whatever you think you'd be able to get working in the time available (and perhaps in the shortest amount of time).

**Comment by ubdbra001** [comment]
There does need to be interaction between the GUI and config as the GUI uses the values in the config file to populate itself, and, when saving, the config should reflect the state of the GUI.

It will still be going through `pyplis/doas_worker` as these will have an instance of the `ConfigHandler` class as an attribute, and then when it comes to saving to a file either will be able to call the saving method in that instance.
We could have it so that the GUI reads and writes its state to the attributes in the `pyplis/doas_worker` instances, and then the `ConfigHandler` class pulls it from there, but that feels like an un-needed step when its possible to access the `ConfigHandler` instance directly through each.

> if we have the GUI interacting too much with the `ConfigHandler` might we end up with a similar situation to what we have now with `PyplisWorker` and the GUI, where we wish they weren't so tangled and are now having to work to disentangle?

It's more about ensuring clean separation of concerns. There needs to be interaction between the front- and back-end, but there shouldn't be any processing steps going on in the front end. The way I'm planning to structure this is that it's just the parameters being passed back and forth, and anything related to saving/loading/checking the config is done in the `ConfigHandler` so it can function independently.
There will be some entanglement between the `PyplisWorker`, `IFitWorker`/`DOASWorker`, and `ConfigHandler` classes, but I think that's necessary.

Always happy to have a chat to explain further, or draw up a diagram to show the interactions.

**Comment by fredsonnenwald** [comment]
If possible I think it would be better try to avoid the use of a global config handler, and instead have a config handler instance inside of pyplis,

**Comment by ubdbra001** [comment]
> each thing that needs configuration owns it's own configuration and has its own config file

The issue with this is that `pyplis_worker` and `doas_worker` currently share a config file, and so need to reference the same object when loading and saving config options. It would be possible to split them I think, but I don't see a huge benefit at this point for the time it would take (plus I'm not sure if it would be possible in the time remaining)

**Comment by twVolc** [comment]
> The issue with this is that `pyplis_worker` and `doas_worker` currently share a config file, and so need to reference the same object when loading and saving config options. It would be possible to split them I think, but I don't see a huge benefit at this point for the time it would take (plus I'm not sure if it would be possible in the time remaining)

Yep, I can see arguments both for and against splitting the attributes out into separate config files. Having them in one makes it easier for the user to just load in a single file and all settings be updated - it might be a bit too easy to forget to load the DOAS config and then have issues recreating a processing procedure. BUT, breaking them into two separate files might actually make things clearer for users and perhaps does make things clearer from a coding perspective too?

It would completely break backwards compatibility, which would be a shame, but maybe at some point these things have to happen anyway, and with the shift to 3.12 coming soon maybe now would be the time to have a big shift if we think it is necessary/makes more sense.

To solve the issue of the user needing to load 2 configs rather than one, the `pyplis_worker` config could hold a file path to the respective `doas_worker` config for that run, so in the GUI, or with headless, you would still only need to select the `pyplis_worker` config and both those settings and doas worker settings could be loaded. If the path is `None` or not there then the `doas_worker` settings just wouldn't be loaded.

This is just thinking out loud and might be adding more complexity for no reason. But it's possible implementing this might not actually be much more work (even could make things easier to code if there's just a base config class witht he specific `pyplis/doas` configs inheriting from this but just changing the attributes they use?). Just a thought anyway.
